integr8ly/cloud-resource-operator
# Added custom metrics for AWS RDS and ElastiCache scheduled maintenance

## Overview

Jira: https://issues.redhat.com/browse/INTLY-4669

## Verification

- Checkout the branch `adding_metrics` on my fork `github.com/davidkirwan/cloud-resource-operator`
- Run `make cluster/prepare`
- Run `make cluster/seed/managed/postgres`
- Run `make cluster/seed/managed/redis`
- Run `make run`
- metrics will be created, `curl localhost:8383/metrics` and expect to see `cro_aws_elasticache_service_maintenance`, `cro_aws_rds_service_maintenance`, `cro_aws_elasticache_info` and `cro_aws_rds_info` eg:
```
cro_aws_elasticache_service_maintenance{AutoUpdateAfterRecommendedApplyByDate="false",Engine="redis",EstimatedUpdateTime="30 minutes per node",ServiceUpdateDescription="Upgrades to improve the security, reliability, and operational performance of your ElastiCache nodes",ServiceUpdateEndDate="1570456320",ServiceUpdateName="elc-20190604-001",ServiceUpdateRecommendedApplyByDate="1563105599",ServiceUpdateReleaseDate="1560513600",ServiceUpdateSeverity="important",ServiceUpdateStatus="expired",ServiceUpdateType="security-update"} 1.563105599e+09
cro_aws_elasticache_service_maintenance{AutoUpdateAfterRecommendedApplyByDate="false",Engine="redis, memcached",EstimatedUpdateTime="30 minutes per node",ServiceUpdateDescription="Upgrades to improve the security, reliability, and operational performance of your ElastiCache nodes",ServiceUpdateEndDate="1581861599",ServiceUpdateName="elc-20191007-001",ServiceUpdateRecommendedApplyByDate="1573826399",ServiceUpdateReleaseDate="1571234400",ServiceUpdateSeverity="important",ServiceUpdateStatus="available",ServiceUpdateType="security-update"} 1.573826399e+09
```

*Note* maintenance metrics might not show, but this is expected if no maintenance is due :)

For verification purposes you can set the `pkg/resources/custom_metrics.go:L13` constant `sleepytime` to `30` seconds, then look for the following log message:

```
INFO[0030] calling reset on all prometheus gauge vectors  custom_metrics=StartGaugeVector
```
